### DISCUSSION TITLE
Enabling  AESGMAC as Authenticated Encryption (AEAD) Algorithms

### DISCUSSION THREAD

**Discussion by vnpraveendesu** [strongswan/strongswan#1005] [Q&A]
Hi 

  I have installed strongSwan ( version 5.9.5) in ubuntu 20.,4 
  I have below queries related to strongSwan configuration
   a. is it possible to use AESGCM as Authenticated Encryption (AEAD) Algorithms without any integrity algorithms
   b. how to enable AESGMAC as  AEAD algorithm Since getting the below error

> failed to create ESP context: unsupported AEAD algorithm NULL_AES_GMAC
> [ESP] failed to create SAD entry
> [ESP] failed to create ESP context: unsupported AEAD algorithm NULL_AES_GMAC
> [ESP] failed to create SAD entry
> [IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel

> List of registered IKE algorithms:
> 
>   encryption: AES_CBC[aes] AES_ECB[aes] 3DES_CBC[des] DES_CBC[des] DES_ECB[des] RC2_CBC[rc2] AES_CFB[openssl]
>               CAMELLIA_CBC[openssl] CAST_CBC[openssl] BLOWFISH_CBC[openssl] NULL[openssl]
>   integrity:  HMAC_MD5_96[openssl] HMAC_MD5_128[openssl] HMAC_SHA1_96[openssl] HMAC_SHA1_128[openssl]
>               HMAC_SHA1_160[openssl] HMAC_SHA2_256_128[openssl] HMAC_SHA2_256_256[openssl] HMAC_SHA2_384_192[openssl]
>               HMAC_SHA2_384_384[openssl] HMAC_SHA2_512_256[openssl] HMAC_SHA2_512_512[openssl] CAMELLIA_XCBC_96[xcbc]
>               AES_XCBC_96[xcbc] AES_CMAC_96[cmac]
>   aead:       AES_GCM_16[openssl] AES_GCM_12[openssl] AES_GCM_8[openssl] AES_CCM_16[openssl] AES_CCM_12[openssl]
>               AES_CCM_8[openssl] CHACHA20_POLY1305[openssl]
>   hasher:     HASH_SHA1[sha1] HASH_SHA2_256[sha2] HASH_SHA2_512[sha2] HASH_SHA2_224[sha2] HASH_SHA2_384[sha2]
>               HASH_SHA3_224[openssl] HASH_SHA3_256[openssl] HASH_SHA3_384[openssl] HASH_SHA3_512[openssl] HASH_MD5[md5]
>               HASH_MD4[openssl] HASH_IDENTITY[openssl]
>   prf:        PRF_KEYED_SHA1[sha1] PRF_HMAC_MD5[openssl] PRF_HMAC_SHA1[openssl] PRF_HMAC_SHA2_256[openssl]
>               PRF_HMAC_SHA2_384[openssl] PRF_HMAC_SHA2_512[openssl] PRF_FIPS_SHA1_160[fips-prf] PRF_AES128_XCBC[xcbc]
>               PRF_CAMELLIA128_XCBC[xcbc] PRF_AES128_CMAC[cmac]
>   xof:        XOF_MGF1_SHA1[mgf1] XOF_MGF1_SHA256[mgf1] XOF_MGF1_SHA512[mgf1] XOF_MGF1_SHA224[mgf1]
>               XOF_MGF1_SHA384[mgf1] XOF_SHAKE128[openssl] XOF_SHAKE256[openssl]
>   drbg:       DRBG_CTR_AES128[drbg] DRBG_CTR_AES192[drbg] DRBG_CTR_AES256[drbg] DRBG_HMAC_SHA1[drbg]
>               DRBG_HMAC_SHA256[drbg] DRBG_HMAC_SHA384[drbg] DRBG_HMAC_SHA512[drbg]
>   dh-group:   MODP_3072[gmp] MODP_4096[gmp] MODP_6144[gmp] MODP_8192[gmp] MODP_2048[gmp] MODP_2048_224[gmp]
>               MODP_2048_256[gmp] MODP_1536[gmp] MODP_1024[gmp] MODP_1024_160[gmp] MODP_768[gmp] MODP_CUSTOM[gmp]
>               CURVE_25519[openssl] CURVE_448[openssl] ECP_256[openssl] ECP_384[openssl] ECP_521[openssl]
>               ECP_224[openssl] ECP_192[openssl] ECP_256_BP[openssl] ECP_384_BP[openssl] ECP_512_BP[openssl]
>               ECP_224_BP[openssl]
>   random-gen: RNG_WEAK[openssl] RNG_STRONG[random] RNG_TRUE[random]
>   nonce-gen:  [nonce]

_

**Comment by tobiasbrunner** [comment]
>  a. is it possible to use AESGCM as Authenticated Encryption (AEAD) Algorithms without any integrity algorithms

Obviously not. The whole point of AEAD algorithms is to efficiently combine encryption and integrity protection.

>  b. how to enable AESGMAC as AEAD algorithm Since getting the below error

You're using the [_kernel-libipsec_ plugin](https://docs.strongswan.org/strongswan-docs/5.9/plugins/kernel-libipsec.html), which implements IPsec in userland and relies on strongSwan plugins to provide the negotiated algorithms. There currently isn't one that implements AES-GMAC. If you want to use it, use the kernel's IPsec implementation via _kernel-netlink_ plugin instead (you might want to do that anyway, read the linked document).
